### PR TITLE
Support secure channels through SSL/TLS

### PR DESCRIPTION
**Pull request by mnito**
`grpclib` does not currently work with secure gRPC channels through HTTPS which we needed for our use case at Slyce -- this pull request uses hyper-h2's SSL context setup to create a secure channel if a secure keyword argument is provided when instantiating the channel.

### CODE REVIEW

File: grpclib/https.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+    # RFC 7540 Section 9.2: Implementations of HTTP/2 MUST use TLS version 1.2
+    # or higher. Disable TLS 1.1 and lower.
+    ctx.options |= (
+        ssl.OP_NO_SSLv2 | ssl.OP_NO_SSLv3 | ssl.OP_NO_TLSv1 | ssl.OP_NO_TLSv1_1
+    )
+
+    # RFC 7540 Section 9.2.1: A deployment of HTTP/2 over TLS 1.2 MUST disable
+    # compression.
+    ctx.options |= ssl.OP_NO_COMPRESSION
```
Comment (vmagamedov) [inline grpclib/https.py]: See above

File: grpclib/https.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+
+def get_http2_ssl_context():
```
Comment (vmagamedov) [inline grpclib/https.py]: I think that it is too early to create additional module just for this function, which will be used only in one place. You can place it in `utils.py` or in `client.py` or directly in the `Channel` class as private method or in `Channel.__init__` method in-place. And I think that there is no need to copy-paste whole function as is with full copyright notice, I think that this code is provided as documentation for reference and it is OK to use it's parts just with a link, pointing to the hyper-h2 project.

**Comment by vmagamedov** [comment]
And please ensure that `:scheme -> https` is used for requests over secure connections.

**Comment by vmagamedov** [comment]
@mnito, will you have a time to finish this PR? Or maybe I can do this for you?

**Comment by mnito** [comment]
Hey @vmagamedov, sorry I haven't gotten to this -- I have some time right now to make the requested changes.

File: grpclib/https.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+    # RFC 7540 Section 9.2.2: "deployments of HTTP/2 that use TLS 1.2 MUST
+    # support TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256". In practice, the
+    # blacklist defined in this section allows only the AES GCM and ChaCha20
+    # cipher suites with ephemeral key negotiation.
+    ctx.set_ciphers("ECDHE+AESGCM:ECDHE+CHACHA20:DHE+AESGCM:DHE+CHACHA20")
+
+    # We want to negotiate using NPN and ALPN. ALPN is mandatory, but NPN may
+    # be absent, so allow that. This setup allows for negotiation of HTTP/1.1.
+    ctx.set_alpn_protocols(["h2", "http/1.1"])
```
[the remark this answers is not in the document]
Comment (mnito) [reply grpclib/https.py]: Is there a significant performance hit or disadvantage to negotiating? Our server requires negotiating and it seems friendlier to set these by default than requiring passing a custom context in situations where negotiating is required.

**Comment by mnito** [comment]
Hey @vmagamedov I made the requested changes and tested with both SSL-enabled and non-SSL-enabled servers. I left the protocol negotiation stuff in because we never had any problems using grpcio with our server and communication did not work without it. If the cost outweighs the benefits of leaving it in, feel free to take it out and we can always pass in a custom SSL context -- but I figured it would be a friendlier default.

File: grpclib/client.py [hunk 4]
```diff
@@ -389,16 +395,39 @@ def __init__(self, host='127.0.0.1', port=50051, *, loop, codec=None):
...
 
+    # https://python-hyper.org/projects/h2/en/stable/negotiating-http2.html
+    def _get_default_ssl_context(self):
+        if not ssl:
+            raise RuntimeError('SSL is not supported.')
+
+        ctx = ssl.create_default_context(purpose=ssl.Purpose.SERVER_AUTH)
+        ctx.options |= (ssl.OP_NO_TLSv1 | ssl.OP_NO_TLSv1_1)
+        ctx.set_ciphers('ECDHE+AESGCM:ECDHE+CHACHA20:DHE+AESGCM:DHE+CHACHA20')
+        ctx.set_alpn_protocols(['h2', 'http/1.1'])
```
Comment (vmagamedov) [inline grpclib/client.py]: I'm totally Ok with this, I don't worry about possible overhead, as this happens only once for a long-living connection. The only concern is that I don't think that we should advertise `http/1.1` protocol support here. Can you remove it from the list and test with your server?

File: grpclib/client.py [hunk 5]
```diff
@@ -389,16 +395,39 @@ def __init__(self, host='127.0.0.1', port=50051, *, loop, codec=None):
                                        header_encoding='utf-8')
         self._authority = '{}:{}'.format(self._host, self._port)
 
+        if ssl and isinstance(ssl, bool):
```
Comment (vmagamedov) [inline grpclib/client.py]: Maybe just `if ssl is True:`?

Comment (mnito) [reply grpclib/client.py]: Doh - yeah

File: grpclib/client.py (hunk 4 above)
Comment (mnito) [reply grpclib/client.py]: Yeah you're probably right -- let me just do a quick test with our server

**Comment by mnito** [comment]
Should be good to go.

**Comment by vmagamedov** [comment]
Everything looks great, can you rebase your changes and resolve merge conflicts?

**Comment by mnito** [comment]
Sure, do you want me to squash as well?

**Comment by vmagamedov** [comment]
Yes! It would be great.

**Comment by mnito** [comment]
I rebased! Thanks for all your input.

**Comment by vmagamedov** [comment]
Thank you for your work!

**Pull request merged**